test/integration/cdi/example-driver/app/kubeletplugin.go

```diff
@@ -0,0 +1,151 @@
+	plugin "k8s.io/kubernetes/test/integration/cdi/example-driver/kubeletplugin"
+)
+
+const (
+	deviceName = "exampledevice"
+	cdiVersion = "0.2.0"
+)
+
+type examplePlugin struct {
+	cdiDir                 string
+	driverName             string
+	draAddress             string
+	pluginRegistrationPath string
+}
+
+// newExamplePlugin returns an initialized examplePlugin instance.
+func newExamplePlugin(cdiDir, driverName, draAddress, pluginRegistrationPath string) (*examplePlugin, error) {
+	return &examplePlugin{
+		cdiDir:                 cdiDir,
+		driverName:             driverName,
+		draAddress:             draAddress,
+		pluginRegistrationPath: pluginRegistrationPath,
+	}, nil
+}
+
+// getJsonFilePath returns the absolute path where json file is/should be.
+func (ex *examplePlugin) getJsonFilePath(claimUID string) string {
+	jsonfileName := fmt.Sprintf("%s-%s.json", ex.driverName, claimUID)
+	return filepath.Join(ex.cdiDir, jsonfileName)
+}
+
+// startPlugin starts servers that are necessary for plugins and registers the plugin with kubelet.
+func (ex examplePlugin) startPlugin() error {
+	klog.Infof("Starting kubelet plugin")
+
+	// create CDI directory if not exists
+	if err := os.MkdirAll(ex.cdiDir, os.FileMode(0750)); err != nil {
+		return fmt.Errorf("failed to create directory: %v", err)
+	}
+
+	// create DRA driver socket directory if not exists.
+	if err := os.MkdirAll(filepath.Dir(ex.draAddress), 0750); err != nil {
+		return fmt.Errorf("failed to create DRA driver socket directory: %v", err)
+	}
+
+	// Run a grpc server for the driver, that listens on draAddress
+	if err := plugin.StartNonblockingGrpcServer(ex.draAddress, &ex); err != nil {
+		return fmt.Errorf("failed to start a nonblocking grpc server: %v", err)
+	}
+
+	// Run a registration server and registers plugin with kubelet
+	if err := plugin.StartRegistrar(ex.driverName, ex.draAddress, ex.pluginRegistrationPath); err != nil {
+		return fmt.Errorf("failed to start registrar: %v", err)
+	}
+
+	return nil
+
+}
+
+func (ex *examplePlugin) NodePrepareResource(ctx context.Context, req *drapbv1.NodePrepareResourceRequest) (*drapbv1.NodePrepareResourceResponse, error) {
+
+	kind := ex.driverName + "/" + deviceName
+
+	klog.Infof("NodePrepareResource is called: request: %+v", req)
+
+	klog.Infof("Creating CDI File")
+	// make resourceHandle in the form of map[string]string
+	var decodedResourceHandle map[string]string
+	if err := json.Unmarshal([]byte(req.ResourceHandle), &decodedResourceHandle); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal resourceHandle: %v", err)
+	}
+
+	// create ContainerEdits in device spec and ContainerEdits
+	envs := []string{}
+	for key, val := range decodedResourceHandle {
+		envs = append(envs, key+"="+val)
+	}
+	// create spec
+	spec := specs.Spec{
+		Version: cdiVersion,
+		Kind:    kind,
+		Devices: []specs.Device{{
+			Name: deviceName,
+			ContainerEdits: specs.ContainerEdits{
+				Env: envs,
+			},
+		}},
```

Comment:
As we don't have any devices here can we skip this and set env vars only in the global container edits section below?